lib/tbot/ca_rotation.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tbot
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// See https://github.com/gravitational/teleport/blob/1aa38f4bc56997ba13b26a1ef1b4da7a3a078930/lib/auth/rotate.go#L135
+// for server side details of how a CA rotation takes place.
+//
+// We can leverage the existing renewal system to fetch new certificates and
+// CAs.
+//
+// We need to force a renewal for the following transitions:
+// - Init -> Update Clients: So we can receive a set of certificates issued by
+//   the new CA, and trust both the old and new CA.
+// - Update Clients, Update Servers -> Rollback: So we can receive a set of
+//   certificates issued by the old CA, and stop trusting the new CA.
+// - Update Servers -> Standby: So we can stop trusting the old CA.
+
+type debouncer struct {
```

Comment:
I've gone for a Timer/Mutex based debouncer here. I originally wrote a channel based implementation:

```go
    debounceChan := make(chan *struct{})
    defer close(debounceChan)
    go func() {
        for {
            select {
            case v := <-debounceChan:
                if v == nil {
                    return
                }
                t := time.NewTimer(time.Second * 10)
                for {
                    select {
                    case <-ctx.Done():
                        return
                    case <-t.C:
                        select {
                        case b.reloadChan <- struct{}{}:
                        default:
                            b.log.Debugf("Renewal already queued, ignoring reload request.")
                        }
                    case v := <-debounceChan:
                        if v == nil {
                            return
                        }
                        // swallow other reload msgs whilst we are in the debounce
                        // period.
                    }
                }
            }
        }
    }()
```

But the channel based implementation seemed a bit more verbose and less clear about what it actually did, even if it handles the context cancellation sooner.